### PR TITLE
None terminated inspectable queue

### PR DESCRIPTION
**Pull request by bedux**
Releated to Issue #1643

### CODE REVIEW

File: core/shared/src/main/scala/fs2/concurrent/Queue.scala [hunk 1]
```diff
@@ -454,6 +454,37 @@ trait InspectableQueue[F[_], A] extends Queue[F, A] {
   def getSize: F[Int]
 }
 
+/**
+  * Like [[InspectableQueue]], but allows allows signalling of no further enqueues by enqueueing `None`.
```
Comment (diesalbla) [inline core/shared/src/main/scala/fs2/concurrent/Queue.scala:458]:
```suggestion
  * Like [[InspectableQueue]], but allows signalling of no further enqueues by enqueueing `None`.
```

File: core/shared/src/main/scala/fs2/concurrent/Queue.scala [hunk 2]
```diff
@@ -454,6 +454,37 @@ trait InspectableQueue[F[_], A] extends Queue[F, A] {
...
+  * Optimizes dequeue to minimum possible boxing.
+  */
+trait NoneTerminatedInspectableQueue[F[_], A] extends NoneTerminatedQueue[F, A] {
+
+  /**
+    * Returns the element which would be dequeued next,
```
Comment (diesalbla) [inline core/shared/src/main/scala/fs2/concurrent/Queue.scala:464]:
```suggestion
    * Gets the element that is next to be dequeued,
```

File: core/shared/src/main/scala/fs2/concurrent/Queue.scala [hunk 3]
```diff
@@ -454,6 +454,37 @@ trait InspectableQueue[F[_], A] extends Queue[F, A] {
...
+    * Offsetting enqueues and de-queues may not result in refreshes.
+    *
+    * Finally, note that operations like `dequeue` are optimized to
+    * work on chunks when possible, which will result in faster
+    * decreases in size that one might expect.
+    * More granular updates can be achieved by calling `dequeue1`
+    * repeatedly, but this is less efficient than dequeueing in
+    * batches.
+    */
+  def size: Stream[F, Int]
```
Comment (diesalbla) [inline core/shared/src/main/scala/fs2/concurrent/Queue.scala:482]: Should we use a `Signal` for this?  _If I understand correctly_, the difference between a Signal and a Stream is that the latter has a returns a fixed number of elements, whereas the former can be sampled by the consumer downstreams and only get values at those points.

File: core/shared/src/main/scala/fs2/concurrent/Queue.scala [hunk 4]
```diff
@@ -465,6 +496,24 @@ object InspectableQueue {
...
+        maxSize: Int
+    )(implicit F: Concurrent[F]): G[NoneTerminatedInspectableQueue[F, A]] =
+      forStrategyNoneTerminated[F, (Boolean, SizedQueue[A]), A](
+        PubSub.Strategy.closeDrainFirst(Queue.Strategy.boundedFifo(maxSize))
+      ) { case (_, h) => h.headOption } { case (_, h) => h.size }
+
+    /** Created an unbounded queue terminated by enqueueing `None`. All elements before `None`. */
+    def noneTerminated[F[_], A](implicit
+        F: Concurrent[F]
+    ): G[NoneTerminatedInspectableQueue[F, A]] =
```
Comment (diesalbla) [inline core/shared/src/main/scala/fs2/concurrent/Queue.scala:510]:
```suggestion
    def noneTerminated[F[_]: Concurrent, A]: G[NoneTerminatedInspectableQueue[F, A]] =
```

No explicit use of the `F` parameter.

File: core/shared/src/main/scala/fs2/concurrent/Queue.scala [hunk 5]
```diff
@@ -572,6 +621,113 @@ object InspectableQueue {
...
+              case Right(_) => -1
+            }
+
+          override def dequeue1: F[Option[A]] =
+            pubSub.get(Right(1)).flatMap {
+              case Left(s) =>
+                Sync[F].raiseError(
+                  new Throwable(
+                    s"Inspectable `dequeue1` requires chunk of size 1 with `A` got Left($s)"
+                  )
```
Comment (diesalbla) [inline core/shared/src/main/scala/fs2/concurrent/Queue.scala:689]: Perhaps we should we use specialised subclasses of `Throwable` for the errors we use.

**Comment by mpilquist** [comment]
What's the status on this one?

**Comment by diesalbla** [comment]
Considering that we have removed the implementations of `Queue` from `fs2`, in favour of those of `cats-effect`, and that Fabio Labella recently introduced the `Channel` as an alternative, I think we can close this one was left behind.

**Pull request closed without merging**